### PR TITLE
Modernize users page - stop using bootstrap classes and start using PF4 components instead

### PR DESCRIPTION
**Pull request by KKoukiou**
New design: Screen Shot 2020-08-27 at 13 50 53

### CODE REVIEW

**Comment by ahmedmgh67** [comment]
This looks nice hope it gets merged and the other page get a redesign to be like this

**Comment by KKoukiou** [comment]
> This looks nice hope it gets merged and the other page get a redesign to be like this

Yes, I am working in parallel on the rest as well. I think the majority of the pages will be ready with the new design within next release (next week)

**Review by andreasn — CHANGES_REQUESTED** [review]
Need some left padding here. Much closer to the left edge than the other pages, and compared to the [Patternfly examples](https://www.patternfly.org/v4/documentation/react/demos/primarydetail/primary--detail-full-page)
Screenshot from 2020-08-27 16-14-17

A bit low contrast. Think it needs darker lines and have them go all the way to the edge.
Screenshot from 2020-08-27 16-36-55

So this didn't align before either, so not a regression in that sense, but now it's shifted to the other side. Will look super-clean if lined up, so if you're able to address that in this PR as well, it would be a very nice bonus!
Screenshot from 2020-08-27 16-37-11

The layout has problems if the parent window is less than 600px, but so did the old one, so not a regression, and can be fixed in a separate PR.
Screenshot from 2020-08-27 16-46-24

The listing of authorized keys in the same resolution did regress though. The **+** and **-** buttons used to be visible, but now they get hidden.
Screenshot from 2020-08-27 16-46-52

**Comment by andreasn** [comment]
Captured the (already existing) issue with the form in issue #14522

File: pkg/users/account-details.js [hunk 1]
```diff
@@ -230,119 +231,112 @@ export function AccountDetails({ accounts, groups, shadow, current_user, user })
...
-                                    <div className="account-column-one">
-                                        <div className="checkbox" data-container="body">
-                                            <label>
-                                                <input type="checkbox" id="account-locked"
-                                   disabled={!superuser.allowed || edited_locked != null}
-                                   checked={edited_locked != null ? edited_locked : details.locked}
-                                   onChange={event => change_locked(event.target.checked)} />
-                                                <span>{_("Lock Account")}</span>
-                                            </label>
+                            <CardTitle id="account-title">{title_name}</CardTitle>
```
Comment (garrett) [inline pkg/users/account-details.js:257]: The cart titles should probably have a heading inside.

Also, the actions should come after the titles, not relying on changing the order via CSS. (It's important to have document order for all types of a11y: keyboard navigation, screen reader, etc.)

Comment (garrett) [reply pkg/users/account-details.js:257]: (Same should apply to the other cards.)

**Review by garrett — CHANGES_REQUESTED** [review]
Looks great overall.

I'd like the cards changed to use a heading and also to use the document order for the actions (so they come after the title, instead of before).

**Comment by garrett** [comment]
Additionally, we should adjust the lists at a top-level instead of each page. The SSH keys list doesn't have a border at the top, it has a border at the bottom, and there's padding around it that shouldn't be there. We keep running into the same things and have only fixed it at the page so far.

File: pkg/users/authorized-keys-panel.js [hunk 2]
```diff
@@ -97,66 +97,71 @@ export function AuthorizedKeys({ name, home, allow_mods }) {
...
     } else
         return null;
 
     return (
-        <div className="panel panel-default" id="account-authorized-keys">
-            <div className="panel-heading">
-                <div className="pull-right">
+        <div className="pf-c-card" id="account-authorized-keys">
+            <div className="pf-c-card__header">
+                <div className="pf-c-card__actions">
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/users/authorized-keys-panel.js:151]: I copied the example in the DOCS which has it in this order, https://www.patternfly.org/v4/documentation/react/components/card#card-header-in-card-head

I understand what you say though, I will adjust our code and send a patch to PF as well.

**Comment by garrett** [comment]
I squashed in a bunch of layout and small size fixes, so the page now wraps properly in smaller widths (narrow desktop, graphical VM @ 1024, mobile), force the key hashes to wrap when there's not enough space, and let the actions wrap under the title.

**Comment by garrett** [comment]
BTW: I think I've addressed all of @andreasn's issues with my changes. Please re-test to be sure. Thanks!

**Review by andreasn — DISMISSED** [review]
Thanks! Much better now!
I think my only last two concerns are:
1. The key empty state lacks some padding. Might also want to place it in the center, since it's a placeholder. but left-aligned works too.
Screenshot from 2020-09-02 18-31-06

2. It seems the SHA-columns doesn't line up, like it does in master, but are dependant on the length of the key name.
Screenshot from 2020-09-02 18-44-22

**Comment by marusak** [comment]
Looks great UI and code wise, thanks!

One issue I noticed is that the password strength is not filled when creating a new account (those 4 bars). They are used when changing password. Not sure if regression or broken on master.
notfilled

**Review by marusak — DISMISSED** [review]
Ok, happens on master as well. Reported as https://github.com/cockpit-project/cockpit/issues/14558
Approving, there are two issues from Andreas, but it might be for followup as it likely comes from a library.

**Comment by KKoukiou** [comment]
> Looks great UI and code wise, thanks!
> 
> One issue I noticed is that the password strength is not filled when creating a new account (those 4 bars). They are used when changing password. Not sure if regression or broken on master.
> notfilled

It's filled just you have to try harder :) When minimum strength there is no line filled.

**Comment by garrett** [comment]
> The key empty state lacks some padding. 

Aaah! I forgot to check a non-me user.

> SHA-columns doesn't line up

This is a flex for some reason, not a table or a grid. It seems PF likes to use non-table `<div>`s to fake tables.

I'm looking into both of these now. Thanks for finding them!

**Comment by marusak** [comment]
> When there's content, it should fill at least one bar. 

Fixed in #14559

**Comment by garrett** [comment]
I've pushed a change that improves the layout for list items and handles the empty state.

For browsers were subgrid is supported and the browser width is wide enough (512px wide in the iframe), I added a grid and brought it down through the children to align the key list cells to the grid, so that they're interdependent and take each other's sizes into account. Currently, only Firefox supports subgrids, so this only applies when you're viewing it in Firefox on the desktop.

For the fallback case and where subgrid is not supported (Chromium and WebKit based browsers as-of this writing), I messed with the flex values, using flex basis, so that it's approximately correct in typical scenarios. Also, this does not affect mobile and very small desktop sizes, where the items still stack.

It's in a separate commit as it's useful to see what I changed and we might want to re-adjust it before landing this PR. (Hopefully it works for everyone and we can just land this. :wink:)

**Review by andreasn — APPROVED** [review]
Subgrid doesn't work on Chrome (yet), but I think it's OK for now. Not too jarring. So good to go!

**Pull request merged**